Encryption.md

```diff
@@ -75,55 +75,76 @@ written to the output stream.
+crypto operations (“AES-NI”) - that can be leveraged, for example, by Java programs 
 (automatically via HotSpot), or C++ programs (via EVP-* functions in OpenSSL).
 
-Initially, two algorithms are implemented, one based on a GCM mode of AES, and the other 
+Initially, two algorithms have been implemented, one based on a GCM mode of AES, and the other 
 on a combination of GCM and CTR modes.
 
 AES-GCM is an authenticated encryption. Besides the data confidentiality (encryption), it 
-supports two levels of integrity verification / authentication: of the data (default), and 
+supports two levels of integrity verification (authentication): of the data (default), and 
 of the data combined with an optional AAD (“additional authenticated data”). The default 
 authentication allows to make sure the data has not been tampered with. An AAD is a free 
 text to be signed, together with the data. The user can, for example, pass the file name 
-with its version (or creation timestamp) as the AAD, to verify the file has not been 
+with its version (or creation timestamp) as the AAD, to verify that the file has not been 
 replaced with an older version.
 
-Sometimes, a hardware acceleration of AES is unavialable (e.g. in Java 8). Then AES crypto 
+Sometimes, a hardware acceleration of AES is unavailable (e.g. in Java 8). Then AES crypto 
 operations are implemented in software, and can be somewhat slow, becoming a performance 
 bottleneck in certain workloads. AES-CTR is a regular (not authenticated) cipher.
 It is faster than AES-GCM, since it doesn’t perform integrity verification and doesn’t 
-calculate the authentication tag. For applications running without AES acceleration and 
+calculate the authentication tag. Actually, AES-GCM is a combination of AES-CTR and an 
+authentication module called GMAC. For applications running without AES acceleration and 
 willing to compromise on content verification, AES-CTR can provide a boost in Parquet 
 write/read throughput. The second Parquet algorithm encrypts the data content (pages) 
 with AES-CTR, and the metadata (Thrift structures) with AES-GCM. This allows to encrypt/decrypt 
 the bulk of the data faster, while still verifying the metadata integrity and making sure 
-the file had not been replaced with a wrong version. However, tampering with the page data 
+the file has not been replaced with a wrong version. However, tampering with the page data 
 might go unnoticed. 
 
-The `AesGcmV1` and `AesGcmCtrV1` structures contain an optional `aad_metadata` field that can 
-be used by a reader to retrieve the AAD string used for file encryption. The maximal allowed
-length of `aad_metadata` is 512 bytes.
-
-Parquet-mr/-cpp implementations use the RBG-based IV construction as defined in the NIST 
-SP 800-38D document for the GCM ciphers (section 8.2.2).
+The initialization vectors (IVs) in AES GCM and CTR ciphers must be unique for each encryption key, 
```

Comment:
Okay, so I just realized that since the IV is stored in the EncryptionAlgorithm and the EncryptionAlgorithm is stored in the FileCryptoMetaData, it means that there is only a single IV for a whole Parquet file. This seems problematic to me, because:
- Either we chain all blocks together, in which case one IV is enough, but we can not decrypt parts of the file without decryipting every block that preceeds it;
- Or we reuse the same IV multiple times, in which case we can selectively decrypt parts of the file, but identical data will lead to identical ciphertexts, weakening the encryption (e.g., if two pages contain the same data).

Do I misunderstand something?

Reply:
Hi Zoltan, have to run now, so will address this comment only, and the others tomorrow morning. We don't keep IVs in the EncryptionAlgorithm. We used to have a so-called iv_prefix for 8.2.1 - but we don't anymore, as per our discussion. Please check the Thrift structures in the doc - iv_prefix'es dont show up there. Once the doc  is signed off, I'll send a pr removing these fields from parquet.thrift.
In any case - we have a fresh IV/nonce for each ciphertext (structure or page) in a file.

Reply:
Great, thanks for the clarification. Just noticed the sentence "The IVs are written before each ciphertext.", sorry for not noticing it earler. Maybe you could make this a bit more explicit by saying that a new, unique IV is generated for every ciphertext and maybe also mention exactly how they are generated (using random number generation).

Reply:
You could also add some diagrams to the corresponding sections like so:

IV (12 bytes) | ciphertext | authentication tag (16 bytes)
------------ | ------------- | -------------

and so:

nonce (16 bytes) | ciphertext
------------ | -------------

Oh, and one more important thing: please explcitily specify the byte order of the raw byte sequences.

Reply:
Zoltan is right. This states that the IV is unique for each key: "The initialization vectors (IVs) in AES GCM and CTR ciphers must be unique for each encryption key, since a single re-use of an IV breaks the encryption."

An IV must be unique for each encrypted stream. The key stream for each IV is identical, so I can take two encrypted streams with the same IV and XOR them to get the plain-text XORed together. If I can control or predict one of the plain-text streams, then I can decrypt without the key.